Enable convolution_backward with bias and symints

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85970

Originally by Krovatkin from https://github.com/pytorch/pytorch/pull/85816

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Enable convolution_backward with bias and symints

Originally by Krovatkin from https://github.com/pytorch/pytorch/pull/85816

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -781,6 +781,19 @@ def stride_expr(name: str) -> str:
+        ),
+        # replace self->sym_sizes() with self_sym_sizes_opt
+        (
+            r"{}->sym_sizes\(\)",
```
Wait, we already replace sym_sizes() into self_sym_sizes two rules above :p 
One of the two needs to be removed.

albanD replied:
Ho that is subtle, my bad!

albanD commented on torchgen/api/translate.py:
```diff
@@ -350,7 +351,16 @@ def direct_solve(goal: NamedCType) -> str:
-            return direct_solve(NamedCType(goal.name, optionalLongVec_ctype))
+            try:
+                argname = direct_solve(NamedCType(goal.name, BaseCType(optionalSymIntArrayRefT)))
+                return f"{argname}.has_value() ? c10::make_optional(c10::asIntArrayRefSlow(*{argname})) : c10::nullopt"
```
Why do we try to generate a SymIntArrayRef first?

albanD commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -8,6 +8,9 @@
 #include <ATen/native/xnnpack/Engine.h>
 #include <ATen/NativeFunctions.h>
 #include <c10/util/accumulate.h>
+#include <ATen/core/ATen_fwd.h>
```
Why these new includes?

albanD reviewed (approved):
Sounds good then.
The translate try/except should do the same thing as before in the worst case and we don't care about codegen speed

ezyang committed:
Update on "Enable convolution_backward with bias and symints"


Originally by Krovatkin from https://github.com/pytorch/pytorch/pull/85816

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
Need to also update the ivalue_to_arg logic

ezyang committed:
Update on "Enable convolution_backward with bias and symints"


Originally by Krovatkin from https://github.com/pytorch/pytorch/pull/85816

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "skip waiting gpu queue"
